### PR TITLE
Add Navigation Bar Implemenation

### LINKED ISSUE #39: Navbar
## Story
As any user, I want to use the navbar to navigate the page.

## Verification
- Verify that the navbar is implemented according to the design.
Bildschirmfoto 2022-04-21 um 08 14 44

### PR DESCRIPTION
**Pull request by simonschuhmacher**
## Summary

- Adds a responsive navigation bar
  - Primary items show current route hierarchy (breadcrumbs)
  - Secondary (profile) items show other configurable links
- Navigation items can be scrolled on mobile devices if overflow happens
- Fixed animation for dropdown component

## Related Issues

- Closes #39 

## Definition of Done

- User story
  - [x] All stated conditions fulfilled
- Code
  - [x] No more code needed
  - [x] No known bugs
- Clean Code
  - [x] Documentation for functions containing logic and components has been added
- Testing
  - [x] All existing tests pass
  - [x] New unit tests created according to [CONTRIBUTING.md](./CONTRIBUTING.md#-testing-strategy)
  - [x] New unit tests pass
- SonarCloud
  - [x] Quality gate passed
- Compatibility to backend
  - [x] Compatibility to staging-backend ([https://saunah-backend-staging.k8s.init-lab.ch](https://saunah-backend-staging.k8s.init-lab.ch)) verified
- Not behind `main`
  - [x] This branch is not behind `main` branch, (current `main` is merged into this branch)

### CODE REVIEW

File: src/components/structural/AppMenu.tsx [hunk 1]
```diff
@@ -1,21 +1,177 @@
...
+export type AppMenuIconItem = {
+    icon: (props: SVGProps<SVGSVGElement>) => JSX.Element
+    iconClasses?: string
+} & AppMenuItemBase
+
+export function isAppMenuIconItem(item: AppMenuItem): item is AppMenuIconItem {
+    return typeof (item as AppMenuIconItem).icon === 'object'
+}
+
+function MenuElement(props: AppMenuItemProps): JSX.Element {
```
Comment (triumphvalley) [inline src/components/structural/AppMenu.tsx:69]: Please add comments for all subcomponents, so it is clear what they do

File: src/components/structural/AppMenu.tsx [hunk 2]
```diff
@@ -1,21 +1,177 @@
...
+                    </Fragment>
+                ))}
+        </>
+    )
+}
+
+function SecondaryMenu({ trailingItem, secondaryItems }: AppMenuProps) {
+    return (
+        <div className="text-right">
+            <Menu as="div" className="relative block text-left">
```
Comment (triumphvalley) [inline src/components/structural/AppMenu.tsx:130]: Couldn't we use the Dropdown here with adjusted styling?

File: src/components/structural/AppMenu.tsx [hunk 3]
```diff
@@ -1,21 +1,177 @@
...
+
+function MenuElement(props: AppMenuItemProps): JSX.Element {
+    return (
+        <OuterAppMenuElement {...props}>
+            <InnerAppMenuElement {...props} />
+        </OuterAppMenuElement>
+    )
+}
+
+type AppMenuItemProps = {
```
Comment (triumphvalley) [inline src/components/structural/AppMenu.tsx:77]: Please rename to corresponding component (I think `MenuElementProps` ?)

File: src/components/structural/AppMenu.tsx [hunk 4]
```diff
@@ -1,21 +1,177 @@
...
+    fromItem?: AppMenuItem
+}
+
+type OuterAppMenuItemProps = {
```
Comment (triumphvalley) [inline src/components/structural/AppMenu.tsx:81]: Please rename to corresponding component.

File: src/components/structural/AppMenu.tsx [hunk 5]
```diff
@@ -1,21 +1,177 @@
...
+export type AppMenuItem = AppMenuTextItem | AppMenuIconItem
+
+type AppMenuItemBase = {
+    url?: string
+    onClick?: () => void
+    testId?: string
+}
+
+export type AppMenuTextItem = {
+    title: ReactElement
```
Comment (triumphvalley) [inline src/components/structural/AppMenu.tsx:53]: Could you make this ReactElement | string? Then we don't have to do the `<>Just a string</>` in the NavigationBar. You could just wrap it inside `InnerAppMenuElement` with `<>fromItem.title</>`

File: src/components/structural/AppMenu.tsx [hunk 6]
```diff
@@ -1,21 +1,177 @@
...
+function MenuElement(props: AppMenuItemProps): JSX.Element {
+    return (
+        <OuterAppMenuElement {...props}>
+            <InnerAppMenuElement {...props} />
+        </OuterAppMenuElement>
+    )
+}
+
+type AppMenuItemProps = {
+    fromItem?: AppMenuItem
```
Comment (triumphvalley) [inline src/components/structural/AppMenu.tsx:78]: Couldn't this be non-optional? Is there a case where a menu element does not have an item?

File: src/views/shared/SaunaDetailBreadcrumb.tsx [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+import api from '../../networking/api'
+
+/**
+ * Dynamic breadcrumb which display the name of the sauna in the
+ * navigation stack. To get the name, it is fetched by id from the
+ * API.
+ */
+function SaunaDetailBreadcrumb({ match, testId }: SaunaDetailBreadcrumbProps<string>): ReactElement {
+    const saunaId = Number(match.params['saunaId'])
+    const [sauna, setSauna] = useState<Sauna.Request>(Sauna.emptyRequest())
```
Comment (triumphvalley) [inline src/views/shared/SaunaDetailBreadcrumb.tsx:13]: I would use `Sauna.Response` here with an undefined initial value. You could set the default value directly in the span below like `sauna?.name || 'Details'`.

File: src/utils/deferred.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
+/**
+ * Utlitity to return a deferred promise, which can
+ * be resolved from outside. Inteded for testing.
+ * @returns promise, resolve and reject
+ */
+export function deferred<T>() {
```
Comment (triumphvalley) [inline src/utils/deferred.ts:6]: noice

File: src/views/shared/NavigationBar.tsx [hunk 9]
```diff
@@ -8,15 +12,39 @@ import AppMenu from '../../components/structural/AppMenu'
...
+    }))
+}
+
+function createSecondaryItems({ isAuthenticated }: AuthState): AppMenuTextItem[] {
+    return [
+        { title: <>Showroom</>, url: '/showroom', testId: 'showroom-test-id' },
+        ...(isAuthenticated
+            ? [
+                  { title: <>Erstellen</>, url: '/saunas/create' },
+                  { title: <>Logout</>, url: '/logout' },
```
Comment (triumphvalley) [inline src/views/shared/NavigationBar.tsx:43]: you could just call logout() here

**Review by triumphvalley — CHANGES_REQUESTED** [review]
Nice 🍞 menu! Just some code comments:

File: src/components/structural/AppMenu.tsx (hunk 2 above)
Comment (triumphvalley) [reply src/components/structural/AppMenu.tsx:130]: Would be desireable but would require some refactoring. It is too much effort I think

**Comment by triumphvalley** [comment]
@simonschuhmacher Do you think we could extract the `MenuLink` into a seperate component which we can use everywhere in the app for links. (Maybe it could have a `showActive` flag, if it should be higlighted if the route is active). Currently we have an anarchy of different stylings for links.

File: src/views/shared/SaunaDetailBreadcrumb.tsx (hunk 7 above)
Comment (simonschuhmacher) [reply src/views/shared/SaunaDetailBreadcrumb.tsx:13]: Nice

File: src/views/shared/NavigationBar.tsx (hunk 9 above)
Comment (simonschuhmacher) [reply src/views/shared/NavigationBar.tsx:43]: 👍🏼

**Comment by simonschuhmacher** [comment]
> 

Not sure, because we use quite specific styling in here and it's pretty specific to the menu, I guess.
Almost all code is specific styling for the menu (e.g. different when route is active) and almost all code that's not the default `Link` component is specific for the menu. So we would just get one bigger component which is harder to read...

File: src/components/structural/AppMenu.tsx (hunk 5 above)
Comment (simonschuhmacher) [reply src/components/structural/AppMenu.tsx:53]: good point

File: src/components/structural/AppMenu.tsx (hunk 4 above)
Comment (simonschuhmacher) [reply src/components/structural/AppMenu.tsx:81]: 👍🏼

File: src/components/structural/AppMenu.tsx (hunk 3 above)
Comment (simonschuhmacher) [reply src/components/structural/AppMenu.tsx:77]: 👍🏼

File: src/components/structural/AppMenu.tsx (hunk 1 above)
Comment (simonschuhmacher) [reply src/components/structural/AppMenu.tsx:69]: 👍🏼

**Review by triumphvalley — APPROVED** [review]
Sehr fein glöst 
Lg G. Gr.

**Pull request merged**